syntax = "proto3";


enum DirectionProto {
    UP = 0;
    DOWN = 1;
    LEFT = 2;
    RIGHT = 3;
}

enum MessageTypeProto {
    CONNECT = 0;
    MOVEMENT = 1;
    DISCONNECT = 2;
}

enum ClientTypeProto {
    LIZARD = 0;
    ROACH = 1;
    WASP = 2;
    DISPLAY_APP = 3;
    LIZARD_BODY = 4;
}

// Message structure
message MessageToServerProto {
    int32 client_id = 1;          // 1 = lizard, 2 = roach, 3 = display-app
    MessageTypeProto type = 2;         // 1 = connect, 2 = movement
    int32 value = 3;              // usage depends on client_id and type
    DirectionProto direction = 4;      // direction to move the roach
    bool message_accepted = 5;    // true = message accepted, false = message rejected
}

// New structures
message LizardProto {
    int32 ch = 1;
    int32 x = 2;
    int32 y = 3;
    int32 score = 4;
    DirectionProto previous_direction = 5;
    bool is_winner = 6;                // True if winner, False otherwise
}

message RoachProto {
    int32 ch = 1;   
    int32 x = 2;
    int32 y = 3;
    bool is_eaten = 4;
    int64 timestamp = 5;
}

message WaspProto {
    int32 ch = 1;
    int32 x = 2;
    int32 y = 3;
}


message FieldUpdateMovementProto {
    int32 num_roaches = 1;
    int32 num_lizards = 2;
    MessageToServerProto message = 3;
    int32 new_x = 4;
    int32 new_y = 5;
    DirectionProto prev_direction = 6;
    bool is_eaten = 7;
}

message FieldUpdateConnectProto {
    int32 client_id = 1;
    int32 position_in_array = 2;
    LizardProto connected_lizard = 3;
    RoachProto connected_roach = 4;
    MessageToServerProto message = 5;
}

message FieldUpdateDisconnectProto {
    int32 client_id = 1;
    int32 position_in_array = 2;
    MessageToServerProto message = 3;
}

message RoachMoverMessageProto {
    repeated RoachProto roaches = 1;         // Array of Roach
    int32 num_roaches = 2;              // Number of roaches
    int32 slot_roaches = 3;             // Slot roaches
}

message LizardMoverMessageProto {
    repeated LizardProto lizards = 1;         // Array of lizards
    int32 num_lizards = 2;              // Number of lizards
    int32 slot_lizards = 3;             // Slot lizards
}

message WaspMoverMessageProto {
    repeated WaspProto wasps = 1;         // Array of Wasp
    int32 num_wasps = 2;              // Number of wasps
    int32 slot_wasps = 3;             // Slot roaches
}